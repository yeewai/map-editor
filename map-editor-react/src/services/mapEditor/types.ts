import { structureDefinitionTypes } from 'services/structureDefinitions';
import { worldTypes } from 'services/worlds';

export interface State {
    zoom: number,
    panV: number,
    panH: number,
    buildStructureDefinition?: structureDefinitionTypes.StructureDefinition
    mousePosition?: { x: number, y: number },
    activeWorld?: worldTypes.World
}
