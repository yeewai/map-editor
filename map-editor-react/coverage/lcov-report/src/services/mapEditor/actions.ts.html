<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/services/mapEditor/actions.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../index.html">All files</a> / <a href="index.html">src/services/mapEditor</a> actions.ts
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>9/9</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>9/9</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>9/9</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">10x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">10x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">10x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">10x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">10x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">10x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">10x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">10x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">10x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import { ActionCreator } from 'redux';
import { Action } from 'services/types';
&nbsp;
import { structureDefinitionTypes } from 'services/structureDefinitions';
import { worldTypes } from 'services/worlds';
&nbsp;
// -------------------------------------------------------------
// "Camera" controls
// -------------------------------------------------------------
export const zoomIn: ActionCreator&lt;Action&gt; = () =&gt; ({
    type: 'mapEditor/ZOOMIN'
});
&nbsp;
export const zoomOut: ActionCreator&lt;Action&gt; = () =&gt; ({
    type: 'mapEditor/ZOOMOUT'
});
&nbsp;
export const zoomReset: ActionCreator&lt;Action&gt; = () =&gt; ({
    type: 'mapEditor/ZOOMRESET'
});
&nbsp;
export const panLeft: ActionCreator&lt;Action&gt; = () =&gt; ({
    type: 'mapEditor/PANLEFT'
});
&nbsp;
export const panRight: ActionCreator&lt;Action&gt; = () =&gt; ({
    type: 'mapEditor/PANRIGHT'
});
&nbsp;
export const panUp: ActionCreator&lt;Action&gt; = () =&gt; ({
    type: 'mapEditor/PANUP'
});
&nbsp;
export const panDown: ActionCreator&lt;Action&gt; = () =&gt; ({
    type: 'mapEditor/PANDOWN'
});
&nbsp;
export const panReset: ActionCreator&lt;Action&gt; = () =&gt; ({
    type: 'mapEditor/PANRESET'
});
&nbsp;
// -------------------------------------------------------------
// Editing controls
// -------------------------------------------------------------
export const setStructureDefinition: ActionCreator&lt;Action&gt; = (sd: structureDefinitionTypes.StructureDefinition ) =&gt; ({
    type: 'mapEditor/SET_BUILD_STRUCTUREDEFINITION',
    payload: sd
});
&nbsp;
// Gonna use this with keyboard controls probably
// export const unsetStructureDefinition: ActionCreator&lt;Action&gt; = () =&gt; ({
//     type: 'mapEditor/UNSET_BUILD_STRUCTUREDEFINITION'
// })
&nbsp;
export const setMousePosition: ActionCreator&lt;Action&gt; = ( x: number, y: number ) =&gt; ({
    type: 'mapEditor/SET_MOUSE_POSITION',
    payload: { x, y }
});
&nbsp;
export const mouseClick: ActionCreator&lt;Action&gt; = ( x: number, y: number ) =&gt; ({
    type: 'mapEditor/MOUSE_CLICK',
    payload: { x, y }
});
&nbsp;
&nbsp;
// -------------------------------------------------------------
// Map Editor Active World
// -------------------------------------------------------------
export const setActiveWorld: ActionCreator&lt;Action&gt; = ( world: worldTypes.World ) =&gt; ({
    type: 'mapEditor/SET_ACTIVE',
    payload: world
});
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Feb 21 2018 10:53:31 GMT-0600 (CST)
</div>
</div>
<script src="../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../sorter.js"></script>
</body>
</html>
