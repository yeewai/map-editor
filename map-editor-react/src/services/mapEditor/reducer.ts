import { Action } from 'services/types';
import * as Types from './types';
import { worldTypes } from 'services/worlds';

export const defaultState: Types.State = {
    zoom: 1,
    panV: 0,
    panH: 0,
    buildStructureDefinition: undefined,
    mousePosition: undefined,
    activeWorld: undefined
}

export const ZOOMRATE = 1.5;
export const MAXZOOM = 10;
export const PANRATE = 1;

const mapEditorReducer = (state = defaultState, action: Action ): Types.State => {
    switch(action.type) {
        case 'mapEditor/SET_ACTIVE':
            return { ...state, activeWorld: action.payload };
        case 'mapEditor/ZOOMIN':
            return { ...state, zoom: Math.min(state.zoom * ZOOMRATE, MAXZOOM) };
        case 'mapEditor/ZOOMOUT':
            return { ...state, zoom: state.zoom / ZOOMRATE };
        case 'mapEditor/ZOOMRESET':
            return { ...state, zoom: defaultState.zoom };
        case 'mapEditor/PANUP':
            return {...state, panV: state.panV - PANRATE}
        case 'mapEditor/PANDOWN':
            return {...state, panV: state.panV + PANRATE}
        case 'mapEditor/PANLEFT':
            return {...state, panH: state.panH - PANRATE}
        case 'mapEditor/PANRIGHT':
            return {...state, panH: state.panH + PANRATE}
        case 'mapEditor/PANRESET':
            return {...state, panH: defaultState.panH, panV: defaultState.panV}
        case 'mapEditor/SET_BUILD_STRUCTUREDEFINITION':
            const sd = action.payload === state.buildStructureDefinition
                ? defaultState.buildStructureDefinition
                : action.payload;
            return {...state, buildStructureDefinition: sd}
        // Will probably use this for keyboard controls
        // case 'mapEditor/UNSET_BUILD_STRUCTUREDEFINITION':
        //     return {...state, buildStructureDefinition: defaultState.buildStructureDefinition}
        //
        case 'mapEditor/SET_MOUSE_POSITION':
            return {...state, mousePosition: action.payload }

        case 'mapEditor/MOUSE_CLICK':
            const { mousePosition, buildStructureDefinition } = state;
            if (!(mousePosition && buildStructureDefinition)) { return state; }

            const newStructure: worldTypes.Structure = {
                ...worldTypes.emptyStructure,
                name: "",
                definitionId: buildStructureDefinition.id,
                definition: buildStructureDefinition,
                xposition: mousePosition.x ,
                yposition: mousePosition.y
            }

            return {
                ...state,
                activeWorld: {
                    ...state.activeWorld,
                    structures: [
                        ...state.activeWorld.structures || [],
                        newStructure
                    ]
                }
            };

        default:
            return state;
    }
}


export default mapEditorReducer;
