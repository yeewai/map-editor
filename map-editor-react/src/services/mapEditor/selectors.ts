import { createSelector } from 'reselect';
import _ from 'lodash';
import { StateTree } from 'services/types';
import { structureDefinitionsSelectors, structureDefinitionTypes as sdTypes } from 'services/structureDefinitions';
import { worldTypes } from 'services/worlds';

// -------------------------------------------------------------
// Camera Controls
// -------------------------------------------------------------
export const getZoom = ( state: StateTree ) => state.mapEditor.zoom;
export const getPan = (state: StateTree ) => ({ h: state.mapEditor.panH, v: state.mapEditor.panV });

export const getBuildStructureDefinition = ( state:StateTree ) => state.mapEditor.buildStructureDefinition;
export const getMousePosition = ( state:StateTree ) => state.mapEditor.mousePosition;

// -------------------------------------------------------------
// Map Editor Active World
// -------------------------------------------------------------
export const getActiveWorld = ( state: StateTree ) => state.mapEditor.activeWorld;

export const getActiveWorldWithDefinition = createSelector(
    getActiveWorld, structureDefinitionsSelectors.get,
    ( world: worldTypes.World, sdefinitions: sdTypes.StructureDefinition[] ) => {
        if (!world) { return null; }
        const structures = world.structures && world.structures.map( s => ({
            ...s,
            withDefinition: true,
            definition: sdefinitions.find( sd => sd.id === s.definitionId)
        }))
        return {
            ...world,
            structures
        };
    }
)

export const getBoardFromWorld = ( world: worldTypes.World ): worldTypes.Structure[][] => {
    //Mutable, but I think it's more efficient tbh...
    let board: worldTypes.Structure[][] = _.range(world.width).map( (x, i) => (_.range(world.length).map( (y, j) => (null))));

    if (world.structures) {
        world.structures.forEach( s => {
            board[s.xposition][s.yposition] = s
        })
    }

    return board;
}

export const getUniqueStructuresInWorld = ( world: worldTypes.World): sdTypes.StructureDefinition[] => {
    return (
        world.structures
        ? Array.from(new Set(world.structures.map( s => s.definition )))
        : []
    );
}

export const getActiveWorldWithBoard = createSelector(
    getActiveWorldWithDefinition, structureDefinitionsSelectors.get,
    ( world: worldTypes.World, sdefinitions: sdTypes.StructureDefinition[] ): worldTypes.WorldWithBoard => {
        if (!world) { return null; }

        return {
            ...world,
            board: getBoardFromWorld(world),
            uniqueStructures: getUniqueStructuresInWorld(world)
        }
    }
)
