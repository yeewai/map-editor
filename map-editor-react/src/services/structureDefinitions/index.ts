import reducer from "./reducer";
import * as structureDefinitionsActions from "./actions";
import * as structureDefinitionsSelectors from "./selectors";
import * as structureDefinitionTypes from "./types";

export { structureDefinitionsActions, structureDefinitionsSelectors, structureDefinitionTypes };
export default reducer;
