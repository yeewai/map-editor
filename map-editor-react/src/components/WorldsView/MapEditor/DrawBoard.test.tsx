import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import sinon from 'sinon';

import { DrawBoard, mapDispatchToProps } from './DrawBoard';
import { mapEditorActions } from 'services/mapEditor';

Enzyme.configure({ adapter: new Adapter() });

describe("Drawing the board", () => {
    let sandbox, wrapper, board;

    beforeEach(() => {
        sandbox = sinon.createSandbox();
        board = [
            [1, 2, 3],
            [4, 5, 6],
            [7, 8, 9]
        ]
    });

    afterEach(() => {
        sandbox.restore();
    });

    it ("renders", () => {

        wrapper = shallow(<DrawBoard board={board} />);
        expect(wrapper).toMatchSnapshot();
    });

    it ("doesn't rerender if the board hasn't changed (eg adjusting the camera)", () => {
        const wrapper = shallow(<DrawBoard board={board} />)

        expect(wrapper.instance().shouldComponentUpdate({board})).toBe(false)
        expect(wrapper.instance().shouldComponentUpdate({board: [[1]]})).toBe(true)
    })

    it ("handles mouse enter", () => {
        const spy = sandbox.stub(mapEditorActions, 'setMousePosition');
        const dispatchSpy = sandbox.spy();

        mapDispatchToProps(dispatchSpy).onMouseEnter(1,2);
        expect(spy.called).toBeTruthy();

        sandbox.restore();
    })

    it ("handles mouse click", () => {
        const spy = sandbox.stub(mapEditorActions, 'mouseClick');
        const dispatchSpy = sandbox.spy();

        mapDispatchToProps(dispatchSpy).onMouseClick(3,4);
        expect(spy.called).toBeTruthy();

        sandbox.restore();
    })

});
