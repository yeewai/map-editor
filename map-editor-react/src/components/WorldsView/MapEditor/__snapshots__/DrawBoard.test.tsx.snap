// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawing the board renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DrawBoard
    board={
        Array [
            Array [
              1,
              2,
              3,
            ],
            Array [
              4,
              5,
              6,
            ],
            Array [
              7,
              8,
              9,
            ],
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          <a
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
>
            <DrawTile
                        structureDefinition={undefined}
                        x={0}
                        y={0}
            />
</a>,
          <a
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
>
            <DrawTile
                        structureDefinition={undefined}
                        x={0}
                        y={1}
            />
</a>,
          <a
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
>
            <DrawTile
                        structureDefinition={undefined}
                        x={0}
                        y={2}
            />
</a>,
        ],
        Array [
          <a
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
>
            <DrawTile
                        structureDefinition={undefined}
                        x={1}
                        y={0}
            />
</a>,
          <a
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
>
            <DrawTile
                        structureDefinition={undefined}
                        x={1}
                        y={1}
            />
</a>,
          <a
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
>
            <DrawTile
                        structureDefinition={undefined}
                        x={1}
                        y={2}
            />
</a>,
        ],
        Array [
          <a
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
>
            <DrawTile
                        structureDefinition={undefined}
                        x={2}
                        y={0}
            />
</a>,
          <a
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
>
            <DrawTile
                        structureDefinition={undefined}
                        x={2}
                        y={1}
            />
</a>,
          <a
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
>
            <DrawTile
                        structureDefinition={undefined}
                        x={2}
                        y={2}
            />
</a>,
        ],
      ],
      "className": "tiles",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "x0y0",
        "nodeType": "host",
        "props": Object {
          "children": <DrawTile
            structureDefinition={undefined}
            x={0}
            y={0}
/>,
          "onMouseDown": [Function],
          "onMouseEnter": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "structureDefinition": undefined,
            "x": 0,
            "y": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "a",
      },
      Object {
        "instance": null,
        "key": "x0y1",
        "nodeType": "host",
        "props": Object {
          "children": <DrawTile
            structureDefinition={undefined}
            x={0}
            y={1}
/>,
          "onMouseDown": [Function],
          "onMouseEnter": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "structureDefinition": undefined,
            "x": 0,
            "y": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "a",
      },
      Object {
        "instance": null,
        "key": "x0y2",
        "nodeType": "host",
        "props": Object {
          "children": <DrawTile
            structureDefinition={undefined}
            x={0}
            y={2}
/>,
          "onMouseDown": [Function],
          "onMouseEnter": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "structureDefinition": undefined,
            "x": 0,
            "y": 2,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "a",
      },
      Object {
        "instance": null,
        "key": "x1y0",
        "nodeType": "host",
        "props": Object {
          "children": <DrawTile
            structureDefinition={undefined}
            x={1}
            y={0}
/>,
          "onMouseDown": [Function],
          "onMouseEnter": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "structureDefinition": undefined,
            "x": 1,
            "y": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "a",
      },
      Object {
        "instance": null,
        "key": "x1y1",
        "nodeType": "host",
        "props": Object {
          "children": <DrawTile
            structureDefinition={undefined}
            x={1}
            y={1}
/>,
          "onMouseDown": [Function],
          "onMouseEnter": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "structureDefinition": undefined,
            "x": 1,
            "y": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "a",
      },
      Object {
        "instance": null,
        "key": "x1y2",
        "nodeType": "host",
        "props": Object {
          "children": <DrawTile
            structureDefinition={undefined}
            x={1}
            y={2}
/>,
          "onMouseDown": [Function],
          "onMouseEnter": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "structureDefinition": undefined,
            "x": 1,
            "y": 2,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "a",
      },
      Object {
        "instance": null,
        "key": "x2y0",
        "nodeType": "host",
        "props": Object {
          "children": <DrawTile
            structureDefinition={undefined}
            x={2}
            y={0}
/>,
          "onMouseDown": [Function],
          "onMouseEnter": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "structureDefinition": undefined,
            "x": 2,
            "y": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "a",
      },
      Object {
        "instance": null,
        "key": "x2y1",
        "nodeType": "host",
        "props": Object {
          "children": <DrawTile
            structureDefinition={undefined}
            x={2}
            y={1}
/>,
          "onMouseDown": [Function],
          "onMouseEnter": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "structureDefinition": undefined,
            "x": 2,
            "y": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "a",
      },
      Object {
        "instance": null,
        "key": "x2y2",
        "nodeType": "host",
        "props": Object {
          "children": <DrawTile
            structureDefinition={undefined}
            x={2}
            y={2}
/>,
          "onMouseDown": [Function],
          "onMouseEnter": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "structureDefinition": undefined,
            "x": 2,
            "y": 2,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "a",
      },
    ],
    "type": "g",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            <a
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
>
              <DrawTile
                            structureDefinition={undefined}
                            x={0}
                            y={0}
              />
</a>,
            <a
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
>
              <DrawTile
                            structureDefinition={undefined}
                            x={0}
                            y={1}
              />
</a>,
            <a
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
>
              <DrawTile
                            structureDefinition={undefined}
                            x={0}
                            y={2}
              />
</a>,
          ],
          Array [
            <a
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
>
              <DrawTile
                            structureDefinition={undefined}
                            x={1}
                            y={0}
              />
</a>,
            <a
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
>
              <DrawTile
                            structureDefinition={undefined}
                            x={1}
                            y={1}
              />
</a>,
            <a
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
>
              <DrawTile
                            structureDefinition={undefined}
                            x={1}
                            y={2}
              />
</a>,
          ],
          Array [
            <a
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
>
              <DrawTile
                            structureDefinition={undefined}
                            x={2}
                            y={0}
              />
</a>,
            <a
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
>
              <DrawTile
                            structureDefinition={undefined}
                            x={2}
                            y={1}
              />
</a>,
            <a
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
>
              <DrawTile
                            structureDefinition={undefined}
                            x={2}
                            y={2}
              />
</a>,
          ],
        ],
        "className": "tiles",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "x0y0",
          "nodeType": "host",
          "props": Object {
            "children": <DrawTile
              structureDefinition={undefined}
              x={0}
              y={0}
/>,
            "onMouseDown": [Function],
            "onMouseEnter": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "structureDefinition": undefined,
              "x": 0,
              "y": 0,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "a",
        },
        Object {
          "instance": null,
          "key": "x0y1",
          "nodeType": "host",
          "props": Object {
            "children": <DrawTile
              structureDefinition={undefined}
              x={0}
              y={1}
/>,
            "onMouseDown": [Function],
            "onMouseEnter": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "structureDefinition": undefined,
              "x": 0,
              "y": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "a",
        },
        Object {
          "instance": null,
          "key": "x0y2",
          "nodeType": "host",
          "props": Object {
            "children": <DrawTile
              structureDefinition={undefined}
              x={0}
              y={2}
/>,
            "onMouseDown": [Function],
            "onMouseEnter": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "structureDefinition": undefined,
              "x": 0,
              "y": 2,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "a",
        },
        Object {
          "instance": null,
          "key": "x1y0",
          "nodeType": "host",
          "props": Object {
            "children": <DrawTile
              structureDefinition={undefined}
              x={1}
              y={0}
/>,
            "onMouseDown": [Function],
            "onMouseEnter": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "structureDefinition": undefined,
              "x": 1,
              "y": 0,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "a",
        },
        Object {
          "instance": null,
          "key": "x1y1",
          "nodeType": "host",
          "props": Object {
            "children": <DrawTile
              structureDefinition={undefined}
              x={1}
              y={1}
/>,
            "onMouseDown": [Function],
            "onMouseEnter": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "structureDefinition": undefined,
              "x": 1,
              "y": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "a",
        },
        Object {
          "instance": null,
          "key": "x1y2",
          "nodeType": "host",
          "props": Object {
            "children": <DrawTile
              structureDefinition={undefined}
              x={1}
              y={2}
/>,
            "onMouseDown": [Function],
            "onMouseEnter": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "structureDefinition": undefined,
              "x": 1,
              "y": 2,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "a",
        },
        Object {
          "instance": null,
          "key": "x2y0",
          "nodeType": "host",
          "props": Object {
            "children": <DrawTile
              structureDefinition={undefined}
              x={2}
              y={0}
/>,
            "onMouseDown": [Function],
            "onMouseEnter": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "structureDefinition": undefined,
              "x": 2,
              "y": 0,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "a",
        },
        Object {
          "instance": null,
          "key": "x2y1",
          "nodeType": "host",
          "props": Object {
            "children": <DrawTile
              structureDefinition={undefined}
              x={2}
              y={1}
/>,
            "onMouseDown": [Function],
            "onMouseEnter": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "structureDefinition": undefined,
              "x": 2,
              "y": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "a",
        },
        Object {
          "instance": null,
          "key": "x2y2",
          "nodeType": "host",
          "props": Object {
            "children": <DrawTile
              structureDefinition={undefined}
              x={2}
              y={2}
/>,
            "onMouseDown": [Function],
            "onMouseEnter": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "structureDefinition": undefined,
              "x": 2,
              "y": 2,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "a",
        },
      ],
      "type": "g",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
